shop/src/main/java/ro/msg/learning/shop/Services/ProductServiceImpl.java

```diff
@@ -0,0 +1,121 @@
+            productRepository.deleteById(id);
+            return "Product deleted!";
+        } else {
+            return "Product not found!";
+        }
+    }
+
+    @Override
+    public ProductDTO updateProduct(ProductDTO newProductValues) {
+        ProductDTO productDTO = new ProductDTO();
+
+        Optional<Product> receivedProductFromDatabase = productRepository.findById(newProductValues.getId());
+
+        // if the product is present in the database, update the product's data
+        if (receivedProductFromDatabase.isPresent()) {
+            Product auxProduct = receivedProductFromDatabase.get();
+            Product failSafeProduct = new Product(
+                    auxProduct.getId(),
+                    auxProduct.getName(),
+                    auxProduct.getDescription(),
+                    auxProduct.getPrice(),
+                    auxProduct.getWeight(),
+                    auxProduct.getImageUrl(),
+                    auxProduct.getCategory(),
+                    auxProduct.getSupplier(),
+                    auxProduct.getOrderDetails(),
+                    auxProduct.getStocks()
+            );
+
+            auxProduct.setName(newProductValues.getName());
+            auxProduct.setDescription(newProductValues.getDescription());
+            auxProduct.setPrice(newProductValues.getPrice());
+            auxProduct.setWeight(newProductValues.getWeight());
+            auxProduct.setImageUrl(newProductValues.getImageUrl());
+            auxProduct.setCategory(productCategoryRepository.findByNameEquals(newProductValues.getProductCategoryName()));
+            auxProduct.setSupplier(supplierRepository.findByNameEquals(newProductValues.getProductSupplierName()));
+
+            try {
+                // And save the new data to the database
+                productDTO = productMapper.mapProductToProductDTO(productRepository.save(auxProduct));
+            } catch (DataIntegrityViolationException exceptionType1) {
+                return productMapper.mapProductToProductDTO(failSafeProduct);
+            }
+        }
+
+        return productDTO;
+    }
+
+    @Override
+    public ProductDTO createProduct(ProductDTO productData) {
+        ProductDTO newlyCreatedProduct;
+
+        // Get the required information for the new product from the DTO and from the database
+        Product toBeCreatedProduct = productMapper.mapProductDTOToProduct(productData);
+        toBeCreatedProduct.setCategory(productCategoryRepository.findByNameEquals(productData.getProductCategoryName()));
+        toBeCreatedProduct.setSupplier(supplierRepository.findByNameEquals(productData.getProductSupplierName()));
+
+        try {
+            // add the new product to the database
+            newlyCreatedProduct = productMapper.mapProductToProductDTO(productRepository.save(toBeCreatedProduct));
+        } catch (DataIntegrityViolationException exceptionType1) {
+            return null;
```

Comment:
Handle the exception, don't return null. How do you handle this null in the controller? What is the user seeing if the product wasn't created?